test: fatal exception: Add code description and testcases

yingmingx opened the pull request:
Improve the test case of CPU exception.
Add equivalence classes and input partition testing
when give an integer reason code.

Signed-off-by: Ying ming <mingx.ying@intel.com>

ioannisg commented on tests/kernel/fatal/exception/src/main.c:
```diff
@@ -90,6 +90,38 @@ void entry_cpu_exception(void *p1, void *p2, void *p3)
+	__asm__ volatile ("svc 0");
+#elif defined(CONFIG_CPU_CORTEX_R)
+	__asm__ volatile ("BKPT");
+#elif defined(CONFIG_ARM)
```
I think I had mentioned that in another PR: could you replace CONFIG_ARM with CONFIG_CPU_CORTEX_M here?

yingmingx replied:
@ioannisg Yeah, another PR is #29528. I close it because I deleted that branch accidentally. 
I distinguish CONFIG_CPU_CORTEX_R because the platform qemu_cortex_r5 will fail when using "swi 0", so I use another instruction to make fatal error insted of "swi 0".
When using CPU_CORTEX_M instead of CONFIG_ARM, building platform on mps2_an385 and mps2_521 will fail. The macro in .config file show there is only CONFIG_ARM without CPU_CORTEX_M. So I think CONFIG_ARM is a more common macro to make the compiler build pass.

yingmingx replied:
@ioannisg  Ping

ioannisg replied:
> When using CPU_CORTEX_M instead of CONFIG_ARM, building platform on mps2_an385 and mps2_521 will fail.

How is this possible, actually? AN385 and AN521 should have both Kconfig options defined; if not, there's a serious bug.

> So I think CONFIG_ARM is a more common macro to make the compiler build pass.

CONFIG_ARM is common to all cortex-variants, but you already have cases for ARM64 and CORTEX_R, here. So doing CORTEX_M should have the same effect.

ioannisg replied:
@yingmingx i think it should not fail the build if you switch CONFIG_ARM with CONFIG_CPU_CORTEX_M, unless i am missing something.

galak replied:
Agree with @ioannisg if this fails with it being CONFIG_CPU_CORTEX_M something really strange is going on.

yingmingx replied:
Hi,@ioannisg . I've found the root cause of my code build failure. After I updated the environment using "west update", there are no warning or error any more. I've updated the code using macro CONFIG_CPU_CORTEX_M and your suggestion is make sense I think.

Commit:
test: fatal exception: Add code description and testcases

Improve the test case of CPU exception.
Add equivalence classes and input partition testing
when give an integer reason code.

Signed-off-by: Ying ming <mingx.ying@intel.com>
